### PR TITLE
Add support for libpwpolicy credit to password policy

### PR DESCRIPTION
**Pull request by rcritten**
We already have libpwquality plumbed in. The ask here was to support the quality credits as well.

There are four credits: digits, uppers, lowers and others

These either count complexity towards satisfying the minimum length or set the minimum number of a type allowed.

If any of the four credits is configured then it will cause the password policy plugin to go through the libpwquality.

Reminder that libpwquality sets its own min password length floor that should be overridden already.

If min_classes is set as well then that is still enforced.

Fixes: https://pagure.io/freeipa/issue/9835

## Summary by Sourcery

Add support for libpwquality "credit" options (dcredit, ucredit, lcredit, ocredit) to adjust password complexity requirements in password policies, integrate them into backend enforcement and CLI, update documentation, and include comprehensive tests.

New Features:
- Introduce digit, uppercase, lowercase, and other character credit settings to password policy

Enhancements:
- Integrate the new credit options into libpwquality enforcement and backend policy code
- Enforce a minimum password length of 6 when any pwquality or credit options are used
- Add validation to prevent using minclasses concurrently with character credits

Documentation:
- Document the new credit parameters in the pwpolicy API reference

Tests:
- Add integration tests covering positive, zero, and negative credit behaviors for each character type

Chores:
- Bump API minor version to include credit options

### CODE REVIEW

File: util/ipa_pwd.c [hunk 1]
```diff
@@ -541,7 +533,7 @@ int ipapwd_check_policy(struct ipapwd_policy *policy,
      * because there are a number of checks that don't have knobs
      * so preserve the previous behavior.
      */
-    check_pwquality = policy->max_repeat + policy->max_sequence + policy->dictcheck + policy->usercheck;
```
[the remark this answers is not in the document]
Comment (rcritten) [reply util/ipa_pwd.c:544]: This is a good idea to simplify and centralize the code. This won't remove all but one of the USE_PWQUALITY blocks because it is also used for compilation and it has nothing to do with fallback: there isn't one.
We need to support the original password policy stuff AND the libpwquality. Users are free to mix if they want though IMHO it complicates things.

File: util/ipa_pwd.c [hunk 2]
```diff
@@ -541,7 +533,7 @@ int ipapwd_check_policy(struct ipapwd_policy *policy,
...
+    check_pwquality = policy->max_repeat + policy->max_sequence + policy->dictcheck + policy->usercheck + (policy->dcredit != 0) + (policy->ucredit != 0) + (policy->lcredit != 0) + (policy->ocredit != 0);
```
[the remark this answers is not in the document]
Comment (rcritten) [reply util/ipa_pwd.c]: Yeah, I'll take the later coding suggestion.

File: util/ipa_pwd.c [hunk 3]
```diff
@@ -634,13 +654,17 @@ char * IPAPWD_ERROR_STRINGS[] = {
...
+    "Password contains too few uppers",
+    "Password contains too few lowers",
+    "Password contains too few others"
 };
 
 char * IPAPWD_ERROR_STRING_GENERAL = "Password does not meet the policy requirements";
 
 char * ipapwd_error2string(enum ipapwd_error err) {
-   if (err < 0 || err > IPAPWD_POLICY_PWD_USER) {
+   if (err < 0 || err > IPAPWD_POLICY_PWD_MIN_OTHERS) {
```
[the remark this answers is not in the document]
Comment (rcritten) [reply util/ipa_pwd.c:705]: This is rarely touched so I'm not super worried about it. The pattern is clear.

File: ipatests/test_integration/test_pwpolicy.py [hunk 4]
```diff
@@ -226,6 +243,89 @@ def test_minclasses(self, reset_pwpolicy):
...
+                    if message in result.stdout_text:
+                        break
+                else:
+                    assert False, "%s not in %s" % (
+                        result.stdout_text, ', '.join(msg)
+                    )
+
+        # hint: minlife == 6
+        # dcredit
+        test_policy(('Pass12', 'password1'), ('123', 'passw'),
```
[the remark this answers is not in the document]
Comment (rcritten) [reply ipatests/test_integration/test_pwpolicy.py:298]: There are, AFAIK, no invalid combinations.

libpwquality enforces its own minimum of 6.

File: ipatests/test_integration/test_pwpolicy.py [hunk 5]
```diff
@@ -226,6 +243,89 @@ def test_minclasses(self, reset_pwpolicy):
...
+                    ('Password is too short',), **{'dcredit': 2})
+        test_policy(('Passw0rd1', 'password12'), ('Passwd1', 'password1'),
+                    ('Password does not contain enough character',
+                     'Password contains too few digits',),
+                    **{'dcredit': -2})
+
+        # ucredit
```
[the remark this answers is not in the document]
Comment (rcritten) [reply ipatests/test_integration/test_pwpolicy.py]: The input values are controlled by the pwquality input parameters. While yes a user could manually directly modify the policy in LDAP we do not support that.

**Comment by rcritten** [comment]
[Link](http://freeipa-org-pr-ci.s3-website.eu-central-1.amazonaws.com/jobs/fe4237fa-6d87-11f0-8512-fa163e030f13) to the temp commit passing

File: util/ipa_pwd.c [hunk 6]
```diff
@@ -634,13 +692,17 @@ char * IPAPWD_ERROR_STRINGS[] = {
     "Password contains a monotonic sequence",
     "Password is based on a dictionary word",
     "Password is a palindrone",
-    "Password contains username"
+    "Password contains username",
+    "Password contains too few digits",
+    "Password contains too few uppers",
+    "Password contains too few lowers",
+    "Password contains too few others"
```
Comment (duzda) [inline util/ipa_pwd.c]: I think changing these to:

```
Password contains too few upper characters.
Password contains too few lower characters.
Password contains too few other characters.
```

would make it much easier to read.
Possibly instead of other characters use special characters?

File: util/ipa_pwd.h [hunk 7]
```diff
@@ -63,6 +67,10 @@ struct ipapwd_policy {
     int max_classrepeat;
     int dictcheck;
     int usercheck;
+    int dcredit;
+    int ucredit;
+    int lcredit;
+    int ocredit;
```
Comment (duzda) [inline util/ipa_pwd.h:74]: Maybe add a comment for reference, seeing these in the wild might raise a few questions?

File: ipaserver/plugins/pwpolicy.py [hunk 8]
```diff
@@ -66,6 +66,18 @@
...
 
+The pwquality options are either mutually exclusive, or take
+precedence, over the standard password policy values. In the case
+of minimum password length if any pwquality-based options are used then
+the minimum length must be >= 6.
+
+The "credit" settings are used to adjust password complexity requirements.
+- With a value of 0, the default, the option is ignored.
+- With a positive value each character of that type in the password
+  contributes towards meeting the mininum length requirement.
```
Comment (duzda) [inline ipaserver/plugins/pwpolicy.py:77]: I'd add an example as the positive might be harder to comprehend and requires testing out.

If I'm understanding correctly, if `minlength=6` and `dcredit=1`, then these are valid:
```
abcdef  // length=6
abcd1  // length=6
abc11  // length=7
```

File: ipatests/test_integration/test_pwpolicy.py [hunk 9]
```diff
@@ -20,7 +20,7 @@ class TestPWPolicy(IntegrationTest):
     """
     Test password policy in action.
     """
-    num_replicas = 1
+    num_replicas = 0
```
Comment (duzda) [inline ipatests/test_integration/test_pwpolicy.py]: Why was this changed?

File: ipatests/test_integration/test_pwpolicy.py [hunk 10]
```diff
@@ -369,11 +469,11 @@ def test_graceperiod_not_replicated(self):
...
-        result = tasks.ldapsearch_dm(
-            self.replicas[0], dn, ['passwordgraceusertime',],
-        )
-        # Never been set at all so won't return
-        assert 'passwordgraceusertime' not in result.stdout_text.lower()
+#        result = tasks.ldapsearch_dm(
+#            self.replicas[0], dn, ['passwordgraceusertime',],
+#        )
+#        # Never been set at all so won't return
+#        assert 'passwordgraceusertime' not in result.stdout_text.lower()
```
Comment (duzda) [inline ipatests/test_integration/test_pwpolicy.py]: And these changes?

File: util/ipa_pwd.c (hunk 6 above)
Comment (rcritten) [reply util/ipa_pwd.c]: Yeah that does read better. I'm going to stick with other characters as that is how libpwquality refers to it. I can be swayed if you really want special.

File: util/ipa_pwd.h (hunk 7 above)
Comment (rcritten) [reply util/ipa_pwd.h:74]: I'm not sure what you mean. Do you mean acknowledge that these are for libpwquality?

File: ipaserver/plugins/pwpolicy.py (hunk 8 above)
Comment (rcritten) [reply ipaserver/plugins/pwpolicy.py:77]: In util/t_policy.c there is a test case for this. It is hard to read due to positional arguments but the function is 
```
static void
set_policy(struct ipapwd_policy *policy,
               int min_pwd_length, int min_diff_chars, int max_repeat,
               int max_sequence, int max_class_repeat, int dict_check,
               int user_check, int dcredit, int ucredit, int lcredit,
               int ocredit)
```
The test is:
```
set_policy(&policy, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0);
assert(ipapwd_check_policy(&policy, "ab21", NULL, 0, 0, 0, 0, NULL) == IPAPWD_POLICY_OK);
```
This uses the default `minlength=6` and shows that for each positive option used the minlength requirement is reduced.

Did you also want something in the integration test?

File: ipatests/test_integration/test_pwpolicy.py (hunk 9 above)
Comment (rcritten) [reply ipatests/test_integration/test_pwpolicy.py]: Ah, it's a leftover from local testing. For the credits I didn't need to involve setting up a replica so I disabled it and the code commented out so the tests would pass (or fail) on a single server. It made local testing far easier and faster so I didn't have to re-run it a bunch on PR-CI.

I'll fix up both.

File: util/ipa_pwd.c (hunk 6 above)
Comment (duzda) [reply util/ipa_pwd.c]: I'll leave this up to you, but I've never seen the term other characters, it makes sense in the context of knowing about digits, lowers and uppers, not so much without any context.

File: util/ipa_pwd.h (hunk 7 above)
Comment (duzda) [reply util/ipa_pwd.h:74]: Yeah, something like that, it's up to you where you want to document this, but I feel like if you document something on the Python side, you should refer to it on the C side. Especially in this case, where *credit doesn't tell you anything.

File: ipaserver/plugins/pwpolicy.py (hunk 8 above)
Comment (duzda) [reply ipaserver/plugins/pwpolicy.py:77]: Oh sorry, I did notice the tests, my issue was, I wasn't really sure how to comprehend the associated docs, so the feature clicked with me after reading the test cases, I don't know how uncommon it is, to have examples in the docs for ipa. It also might appear in the web docs, which would be probably more than enough. Just wanted to raise a suggestion, that it's not super clear, if you don't know libpwquality.

```
The pwquality options are either mutually exclusive, or take
precedence, over the standard password policy values. In the case
of minimum password length if any pwquality-based options are used then
the minimum length must be >= 6.

The "credit" settings are used to adjust password complexity requirements.
- With a value of 0, the default, the option is ignored.
- With a positive value each character of that type in the password
  contributes towards meeting the mininum length requirement.

 For a password policy of `minlength=6`, `dcredit=1`, these passwords are valid:
    abcdef
    abcd1

- With a negative value signifies a minimum number of that character type
  that must be present.
```

The tests are fine, no need to touch those!

File: util/ipa_pwd.c (hunk 6 above)
Comment (rcritten) [reply util/ipa_pwd.c]: I'm fine with special.

File: util/ipa_pwd.h (hunk 7 above)
Comment (rcritten) [reply util/ipa_pwd.h:74]: I added a block comment around the libpwquality options. *credit mirrors the option names within libpwquality. I was trying to keep things consistent.

File: ipaserver/plugins/pwpolicy.py (hunk 8 above)
Comment (rcritten) [reply ipaserver/plugins/pwpolicy.py:77]: I added an example with slight format changes. And yes, this is difficult to understand. It is the reason I skipped the *option settings when I first implemented libpwquality support but since I opened the door to them they were naturally requested.

Comment (duzda) [reply ipaserver/plugins/pwpolicy.py:77]: Thank you, I think this will be helpful for the users

File: util/ipa_pwd.h (hunk 7 above)
Comment (duzda) [reply util/ipa_pwd.h:74]: Awesome

**Review by duzda — APPROVED** [review]
LGTM, I think you will have to do a rebase later(?)

**Comment by rcritten** [comment]
master:

* 892fea881a6ceef80affd6d6a3fb9b5afafa969b Add support for libpwpolicy credit to password policy
